lib/core/src/text/content.rs

```diff
@@ -0,0 +1,370 @@
+    pub dirty_lines : DirtyLines,
+    pub font        : &'b mut FontRenderInfo,
+}
+
+impl TextComponentContent {
+    /// Create a text component containing `text`
+    ///
+    /// The text will be split to lines by `'\n'` characters.
+    pub fn new(font_id:FontId, text:&str) -> Self {
+        TextComponentContent {
+            lines       : Self::split_to_lines(text),
+            dirty_lines : DirtyLines::default(),
+            font        : font_id,
+        }
+    }
+
+    fn split_to_lines(text:&str) -> Vec<String> {
+        let split      = text.split('\n');
+        let without_cr = split.map(Self::cut_cr);
+        without_cr.map(|s| s.to_string()).collect()
+    }
+
+    fn cut_cr(from:&str) -> &str {
+        if from.ends_with('\r') {
+            &from[..from.len()-1]
+        } else {
+            from
+        }
+    }
+
+    /// Get RefreshInfo for this content.
+    ///
+    /// The dirty flags for lines are moved to returned content, so the `self` dirty flags will be
+    /// cleared after this call.
+    pub fn refresh_info<'a,'b>(&'a mut self, fonts:&'b mut Fonts) -> RefreshInfo<'a,'b> {
+        RefreshInfo {
+            lines       : &mut self.lines,
+            dirty_lines : std::mem::take(&mut self.dirty_lines),
+            font        : fonts.get_render_info(self.font)
+        }
+    }
+
+    /// Apply change to content.
+    pub fn make_change(&mut self, change:TextChange) {
+        match change.change_type() {
+            ChangeType::Simple    => self.make_simple_change(change),
+            ChangeType::Multiline => self.make_multiline_change(change),
+        }
+    }
+    
+    fn make_simple_change(&mut self, change:TextChange) {
+        let line_index  = change.replaced.start.line;
+        let new_content = change.lines.first().unwrap();
+        let range       = change.replaced.start.byte_offset..change.replaced.end.byte_offset;
+        self.lines[line_index].replace_range(range, new_content);
+        self.dirty_lines.add_single_line(line_index);
+    }
+
+    fn make_multiline_change(&mut self, mut change:TextChange) {
+        let start_line           = change.replaced.start.line;
+        let end_line             = change.replaced.end.line;
+        let replaced_lines_count = end_line - start_line + 1;
+        let inserted_lines_count = change.lines.len();
+
+        self.mix_content_into_change(&mut change);
+        self.lines.splice(start_line..=end_line, change.lines);
+        if replaced_lines_count != inserted_lines_count {
+            self.dirty_lines.add_lines_range_from(start_line..);
+        } else {
+            self.dirty_lines.add_lines_range(start_line..=end_line);
+        }
+    }
+    
+    fn mix_content_into_change(&mut self, change:&mut TextChange) {
+        self.mix_first_edited_line_into_change(change);
+        self.mix_last_edited_line_into_change(change);
+    }
+
+    fn mix_first_edited_line_into_change(&self, change:&mut TextChange) {
```

Comment:
Lacks doucmentation. I know its private, still I dont understand what it really does by the name (what does "mix" mean?)

Reply:
Right, it's not obvious even in context of its usage. I'll add a doc comment.